src/rpc.rs

```diff
@@ -1,103 +1,559 @@
+		self.client.read().await.0.clone()
+	}
+
+	pub async fn current_node(&self) -> Node {
+		self.client.read().await.1.clone()
+	}
+
+	async fn with_client<F, Fut, T>(&self, mut f: F) -> Result<T>
+	where
+		F: FnMut(avail::Client) -> Fut + Copy,
+		Fut: std::future::Future<Output = Result<T, subxt::error::Error>>,
+	{
+		let either = backoff::future::retry(self.backoff.clone(), move || async move {
+			// Try with current client
+			match f(self.current_client().await).await {
+				Ok(ok) => return Ok(Either::Left(ok)),
+				Err(error) => {
+					warn!(%error, "Failed to connect to node. Trying to reach to another one");
+				},
+			}
+			let last_node = self.current_node().await;
+
+			let mut nodes = self.nodes.clone();
+			// Shuffle nodes discarding the last one
+			Self::shuffle_full_nodes(&mut nodes, Some(&last_node.host));
+			// Trying to find some node which is available
+			Self::connect_to_available_rpc_and_return_call(
+				&nodes,
+				&self.expected_version,
+				Some(last_node.genesis_hash),
+				move |cl| f(cl).map_err(anyhow::Error::from),
+			)
+			.await
+			.map_err(backoff::Error::transient)
+			.map(Either::Right)
+		})
 		.await
-		.unwrap();
+		.context("Failed to reach to any node")?;
 
-	// Drop weights, as they are not currently used.
-	Ok(grandpa_valset.iter().map(|e| e.0).collect())
-}
+		let (client, node, ok) = match either {
+			// If current client succeeded we don't need to update DB and structure
+			Either::Left(ok) => return Ok(ok),
+			Either::Right(ret) => ret,
+		};
 
-pub async fn get_valset_by_block_number(
-	client: &avail::Client,
-	block: u32,
-) -> Result<Vec<ed25519::Public>> {
-	let hash = get_block_hash(client, block).await?;
-	get_valset_by_hash(client, hash).await
-}
-/// RPC for obtaining header of latest finalized block mined by network
-pub async fn get_chain_head_header(client: &avail::Client) -> Result<DaHeader> {
-	let h = client.rpc().finalized_head().await?;
-	client
-		.rpc()
-		.header(Some(h))
+		// Update last full node in DB
+		if let Some(db) = &self.db {
+			crate::data::store_last_full_node_ws_in_db(db.clone(), node.host.clone())?;
+		}
+
+		// And in structure
+		*self.client.write().await = (client, node);
+
+		Ok(ok)
+	}
+
+	fn with_client_subscribe<T, F, Fut>(self, f: F) -> impl Stream<Item = anyhow::Result<T>>
+	where
+		F: FnMut(avail::Client) -> Fut + Copy,
+		Fut: std::future::Future<
+			Output = Result<avail_subxt::rpc::Subscription<T>, subxt::error::Error>,
+		>,
+		T: serde::de::DeserializeOwned,
+	{
+		async_stream::stream! {
+			'outer: loop  {
+				let mut stream = match self.with_client(f).await {
+					Ok(s) => s,
+					Err(err) => {
+						yield Err(err);
+						return;
+					}
+				};
+
+				loop {
+					// We need to return, as stream has ended
+					let Some(result) = stream.next().await else { return };
+					// If we received error that means that we need to find a new client
+					let Ok(res) = result else { continue 'outer };
+					yield Ok(res);
+				}
```

Comment:
I've retested everything, and it seems that during the connection drop, instead of sending an error rpc, jsonrpsee subscription sends `None` which is weird.

So I'll revert it back to:
```rust
while let Some(Ok(res)) = stream.next().await {
    yield Ok(res);
}
```

Reply:
Or if we know that some subscriptions are endless, we can make other workaround